ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+  }
+
+  void scheduleHeartbeatRelatedTasks() {
+    SendHeartbeatTask sendTask = new SendHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, configs.heartbeatSendIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    CheckHeartbeatTask processTask = new CheckHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(processTask, 0, configs.heartbeatCheckIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    DiscoverClusterTask discoverTask = new DiscoverClusterTask();
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, configs.discoverClusterIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
+                            final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        aliveHosts.put(host.toString(), false);
```

Comment:
do we need to maintain the not-alive hosts? if not, can we remove them? if you do need the not-alive hosts, it might be good to have a map from host -> lastAliveTime instead of just true/false so that you have meaningful information about when it was last alive (this will also help you if these threads die)

Reply:
I keep the not-alive hosts for being able to observe and debug the cluster. If we don't have it, we have to look at previous responses to infer what servers were part of the cluster at some point in time and see if a server was set to down.
The true/false currently is just a temporary value. In the future, we want to add extra information like the lag, offset and maybe also `lastAliveTime` you suggest.  I can add the `lastAliveTime` but again, by itself it's not meaningful as it doesn't tell me if the server is up or down. I can add it though in addition to true/false

Reply:
> I can add the lastAliveTime but again, by itself it's not meaningful as it doesn't tell me if the server is up or down. I can add it though in addition to true/false

I was actually thinking the opposite! Without `lastAliveTime` you could think a server is up but it hasn't been updated in a long time because the checker task is dead. It might make sense to (wherever you check for is alive flag) instead check for `lastAliveTime > now() - windowSize`. I suppose we could say that it's all optimistic and that we'll handle failures on the callee, but we should call out that assumption

Reply:
May be we need both.. status and a last check time ?